holopy/inference/tests/test_functional_api.py

```diff
@@ -0,0 +1,137 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with HoloPy.  If not, see <http://www.gnu.org/licenses/>.
+
+
+import unittest
+import warnings
+
+import numpy as np
+from nose.plugins.attrib import attr
+
+from holopy.core.metadata import data_grid
+from holopy.scattering import Sphere
+from holopy.inference import sample, fit, prior, ExactModel, EmceeStrategy
+from holopy.inference.functional_api import (
+    make_default_model, parameterize_scatterer, rename_xyz, make_uniform)
+from holopy.inference.result import SamplingResult
+from holopy.inference.tests.common import SimpleModel
+
+DATA = data_grid(np.ones((2, 2)), spacing=1, medium_index=1,
+    illum_wavelen=0.5, illum_polarization = [0,1])
+SPHERE = Sphere(n=1, center=[2, 2, 2])
+GUESSES = {'n': 1, 'r': 2, 'center.0': 3}
+
+class TestUserFacingFunctions(unittest.TestCase):
+    @attr('fast')
+    def test_cannot_sample_without_model(self):
+        self.assertRaises(ValueError, sample, DATA, Sphere())
+
+    @attr('medium')
+    def test_sample_function_samples(self):
+        result = sample(DATA, SimpleModel())
+        self.assertTrue(isinstance(result, SamplingResult))
+        self.assertTrue(isinstance(result.strategy, EmceeStrategy))
+        self.assertTrue(hasattr(result, 'samples'))
+
+    @attr('medium')
+    def test_fit_works_with_model(self):
+        function_result = fit(DATA, SimpleModel())
+        function_result.time = None
+        object_result = SimpleModel().fit(DATA)
+        object_result.time = None
+        self.assertEqual(function_result, object_result)
+
+    @attr('medium')
+    def test_fit_works_with_scatterer(self):
+        function_result = fit(DATA, SPHERE)
+        function_result.time = None
+        model = make_default_model(SPHERE, None)
+        object_result = model.fit(DATA)
+        object_result.time = None
+        self.assertEqual(function_result, object_result)
+
+    @attr('medium')
+    def test_fit_works_with_scatterer_and_parameters(self):
+        function_result = fit(DATA, SPHERE, ['n', 'x'])
+        function_result.time = None
+        model = make_default_model(SPHERE, ['n', 'x'])
+        object_result = model.fit(DATA)
+        object_result.time = None
+        self.assertEqual(function_result, object_result)
+
+    @attr('medium')
+    def test_model_takes_precendence_over_parameters(self):
+        model = make_default_model(SPHERE, ['n', 'x'])
+        with warnings.catch_warnings():
+            warnings.simplefilter('ignore', UserWarning)
+            result = fit(DATA, model, ['r', 'y'])
+        self.assertEqual(result._names, ['n', 'center.0'])
+
+    @attr('medium')
+    def test_passing_model_and_parameters_gives_warning(self):
+        model = make_default_model(SPHERE, None)
+        self.assertWarns(UserWarning, fit, DATA, model, ['r', 'y'])
+
```

Comment:
Can we get a `def test_sample_raises_error_when_invalid_model`? Otherwise I think you hit all the directions of the functional api.

Reply:
That is actually the first test in `test_interface`:

```python
    @attr('fast')
    def test_cannot_sample_without_model(self):
        self.assertRaises(ValueError, sample, DATA, Sphere())
```
unless I am misunderstanding the test you are requesting...

You did inspire me to add a test that the `fit` function doesn't work with arbitrary input.